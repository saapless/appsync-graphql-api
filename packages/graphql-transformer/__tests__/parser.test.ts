import { print } from "graphql";
import {
  DirectiveNode,
  DocumentNode,
  FieldNode,
  InputValueNode,
  InterfaceNode,
  NamedTypeNode,
  NonNullTypeNode,
  ObjectNode,
} from "../src/parser";

const schema = /* GraphQL */ `
  type Viewer

  type User @model {
    id: ID!
    firstName: String
    lastName: String
    email: AWSEmail @auth(rules: [{ allow: "owner" }])
    picture: AWSURL
  }

  type Task @model {
    id: ID!
    title: String!
    content: AWSJSON
  }

  extend type Viewer {
    user: User
    tasks: Task @connection
  }

  type Query {
    hello: String
  }
`;

describe("DocumentNode parser", () => {
  const documentNode = DocumentNode.fromSource(schema);

  it("should parse document node", () => {
    const queryNode = documentNode.getNode("Query");
    expect(queryNode).toBeDefined();
    expect(queryNode).toBeInstanceOf(ObjectNode);
  });

  it("should serialize document", () => {
    const serialized = documentNode.serialize();
    expect(print(serialized)).toMatchSnapshot();
  });

  it("should add nodes", () => {
    const nodeInterface = InterfaceNode.create("Node", [
      FieldNode.create("id", NonNullTypeNode.create(NamedTypeNode.create("ID"))),
    ]);

    documentNode.addNode(nodeInterface);

    const queryNode = documentNode.getNode("Query");

    if (queryNode instanceof ObjectNode) {
      queryNode.addField(
        FieldNode.create(
          "node",
          NamedTypeNode.create("Node"),
          [InputValueNode.create("id", NonNullTypeNode.create(NamedTypeNode.create("ID")))],
          [DirectiveNode.create("auth")]
        )
      );
    }

    expect(print(documentNode.serialize())).toMatchSnapshot();
  });
});
