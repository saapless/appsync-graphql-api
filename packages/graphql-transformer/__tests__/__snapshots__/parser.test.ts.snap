// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentNode parser should add nodes 1`] = `
"type Viewer {
  user: User
  tasks: Task @connection
}

type User @model {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail @auth(rules: [{allow: "owner"}])
  picture: AWSURL
}

type Task @model {
  id: ID!
  title: String!
  content: AWSJSON
}

type Query {
  hello: String
  node(id: ID!): Node @auth
}

interface Node {
  id: ID!
}"
`;

exports[`DocumentNode parser should serialize document 1`] = `
"type Viewer {
  user: User
  tasks: Task @connection
}

type User @model {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail @auth(rules: [{allow: "owner"}])
  picture: AWSURL
}

type Task @model {
  id: ID!
  title: String!
  content: AWSJSON
}

type Query {
  hello: String
}"
`;
